File: ChatApp/app/src/main/java/top/chilfish/chatapp/adapter/BaseAdapter.java
package top.chilfish.chatapp.adapter;

import android.annotation.SuppressLint;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

  private List<T> mDataList;
  private OnItemClickListener<T> mOnItemClickListener;

  public BaseAdapter(List<T> dataList) {
    mDataList = dataList;
  }

  @SuppressLint("NotifyDataSetChanged")
  public void setDataList(List<T> dataList) {
    mDataList = dataList;
    notifyDataSetChanged();
  }

  public List<T> getDataList() {
    return mDataList;
  }

  public T getItem(int position) {
    if (mDataList == null || position < 0 || position >= mDataList.size()) {
      return null;
    }
    return mDataList.get(position);
  }

  @Override
  public int getItemCount() {
    return mDataList == null ? 0 : mDataList.size();
  }

  @Override
  public int getItemViewType(int position) {
    return super.getItemViewType(position);
  }

  protected abstract int getItemLayout();

  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder Holder, int position) {
    ViewHolder holder = (ViewHolder) Holder;
    try {
      T item = getItem(position);
      holder.bindData(item);
      holder.itemView.setOnClickListener(v -> {
        if (mOnItemClickListener != null) {
          mOnItemClickListener.onItemClick(item);
        }
      });
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  public class ViewHolder extends RecyclerView.ViewHolder {
    public ViewHolder(@NonNull View itemView) {
      super(itemView);
    }

    public void bindData(T data) {
    }
  }

  public void setOnItemClickListener(OnItemClickListener<T> listener) {
    mOnItemClickListener = listener;
  }

  public interface OnItemClickListener<T> {
    void onItemClick(T item);
  }
}

File: ChatApp/app/src/main/java/top/chilfish/chatapp/adapter/BaseAdapter.kt
package top.chilfish.chatapp.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter<T, VB : ViewDataBinding> :
    RecyclerView.Adapter<BaseAdapter.ViewHolder<VB>>() {

    protected var items: List<T> = listOf()
    private var onItemClickListener: ((T) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<VB> {
        val inflater = LayoutInflater.from(parent.context)
        val binding = DataBindingUtil.inflate<VB>(inflater, getLayoutId(), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder<VB>, position: Int) {
        val item = items[position]
        holder.bind(item)
        holder.itemView.setOnClickListener {
            onItemClicked(item, holder.itemView)
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun updateItems(newItems: List<T>) {
        items = newItems
        notifyDataSetChanged()
    }

    abstract fun getLayoutId(): Int

    open fun onItemClicked(item: T, view: View) {
        onItemClickListener?.invoke(item)
    }

    class ViewHolder<VB : ViewDataBinding>(val binding: VB) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(item: Any?) {
//            binding.setVariable(BR.item, item)
            binding.executePendingBindings()
        }
    }
}